File: frontend/src/api/planejamento.ts
import { apiFetch } from "./client";
import type { TransacaoTipo } from "./dashboards";

export type CelulaOrigem =
  "realizado" | "sugerido" | "confirmado" | "hipotetico" | "cumprido" | "vazio";

export interface CelulaGrade {
  ano: number;
  mes: number;
  valor: string;
  origem: CelulaOrigem;
  realizado_parcial: string | null;
  status: "dentro" | "alerta" | null;
  transacao_vinculada_id: number | null;
}

export interface LinhaSubcategoriaGrade {
  subcategory_id: number;
  subcategory_nome: string;
  group_id: number;
  group_nome: string;
  tipo: TransacaoTipo;
  celulas: CelulaGrade[];
}

export interface LinhaItemGrade {
  item_id: number;
  nome: string;
  tipo: TransacaoTipo;
  recorrente: boolean;
  cumprido: boolean;
  celulas: CelulaGrade[];
}

export interface LinhaEventualGrade {
  tipo: TransacaoTipo;
  celulas: CelulaGrade[];
}

export interface Grade {
  periodo: [number, number][];
  subcategorias: LinhaSubcategoriaGrade[];
  itens: LinhaItemGrade[];
  eventuais: LinhaEventualGrade[];
  total_despesas: CelulaGrade[];
  total_receitas: CelulaGrade[];
  saldo: CelulaGrade[];
}

export function fetchPlanejamentoGrade(anoBase: number, mesBase: number): Promise<Grade> {
  return apiFetch<Grade>(`/planejamento/grade?ano_base=${anoBase}&mes_base=${mesBase}`);
}

export function confirmarPlanejamentoValor(
  subcategoryId: number,
  input: { ano: number; mes: number; valor: string }
): Promise<unknown> {
  return apiFetch(`/planejamento/valores/${subcategoryId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
}

export function removerPlanejamentoValor(
  subcategoryId: number,
  ano: number,
  mes: number
): Promise<void> {
  return apiFetch<void>(`/planejamento/valores/${subcategoryId}?ano=${ano}&mes=${mes}`, {
    method: "DELETE",
  });
}

export interface ItemPlanejado {
  id: number;
  nome: string;
  tipo: TransacaoTipo;
  valor: string;
  data_inicio: string;
  recorrente: boolean;
  data_fim: string | null;
  transacao_vinculada_id: number | null;
}

export interface ItemPlanejadoInput {
  nome: string;
  tipo: TransacaoTipo;
  valor: string;
  dataInicio: string;
  recorrente: boolean;
  dataFim: string | null;
}

function toBody(input: ItemPlanejadoInput) {
  return {
    nome: input.nome,
    tipo: input.tipo,
    valor: input.valor,
    data_inicio: input.dataInicio,
    recorrente: input.recorrente,
    data_fim: input.dataFim,
  };
}

export function fetchItensPlanejados(): Promise<ItemPlanejado[]> {
  return apiFetch<ItemPlanejado[]>("/planejamento/itens");
}

export function createItemPlanejado(input: ItemPlanejadoInput): Promise<ItemPlanejado> {
  return apiFetch<ItemPlanejado>("/planejamento/itens", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(toBody(input)),
  });
}

export function updateItemPlanejado(
  itemId: number,
  input: ItemPlanejadoInput
): Promise<ItemPlanejado> {
  return apiFetch<ItemPlanejado>(`/planejamento/itens/${itemId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(toBody(input)),
  });
}

export function deleteItemPlanejado(itemId: number): Promise<void> {
  return apiFetch<void>(`/planejamento/itens/${itemId}`, { method: "DELETE" });
}

export function vincularItemPlanejado(itemId: number, transacaoId: number): Promise<ItemPlanejado> {
  return apiFetch<ItemPlanejado>(`/planejamento/itens/${itemId}/vincular`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ transacao_id: transacaoId }),
  });
}

export function desvincularItemPlanejado(itemId: number): Promise<ItemPlanejado> {
  return apiFetch<ItemPlanejado>(`/planejamento/itens/${itemId}/desvincular`, {
    method: "POST",
  });
}

File: backend/app/schemas/planejamento.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, model_validator

from app.models.pluggy import PluggyTransactionTipo


class CelulaGradeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    ano: int
    mes: int
    valor: Decimal
    origem: str
    realizado_parcial: Decimal | None = None
    status: str | None = None
    transacao_vinculada_id: int | None = None


class LinhaSubcategoriaGradeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    subcategory_id: int
    subcategory_nome: str
    group_id: int
    group_nome: str
    tipo: PluggyTransactionTipo
    celulas: list[CelulaGradeOut]


class LinhaItemGradeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    item_id: int
    nome: str
    tipo: PluggyTransactionTipo
    recorrente: bool
    cumprido: bool
    celulas: list[CelulaGradeOut]


class LinhaEventualGradeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    tipo: PluggyTransactionTipo
    celulas: list[CelulaGradeOut]


class GradeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    periodo: list[tuple[int, int]]
    subcategorias: list[LinhaSubcategoriaGradeOut]
    itens: list[LinhaItemGradeOut]
    eventuais: list[LinhaEventualGradeOut]
    total_despesas: list[CelulaGradeOut]
    total_receitas: list[CelulaGradeOut]
    saldo: list[CelulaGradeOut]


class PlanejamentoValorIn(BaseModel):
    ano: int
    mes: int
    valor: Decimal


class PlanejamentoValorOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    subcategory_id: int
    ano: int
    mes: int
    valor: Decimal


class PlanejamentoValorEventualIn(BaseModel):
    ano: int
    mes: int
    valor: Decimal


class PlanejamentoValorEventualOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tipo: PluggyTransactionTipo
    ano: int
    mes: int
    valor: Decimal


class ItemPlanejadoIn(BaseModel):
    nome: str
    tipo: PluggyTransactionTipo
    valor: Decimal
    data_inicio: date
    recorrente: bool = False
    data_fim: date | None = None

    @model_validator(mode="after")
    def _valida_data_fim(self) -> "ItemPlanejadoIn":
        if self.data_fim is not None and not self.recorrente:
            raise ValueError("data_fim só é válido para item planejado recorrente")
        return self


class ItemPlanejadoOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    nome: str
    tipo: PluggyTransactionTipo
    valor: Decimal
    data_inicio: date
    recorrente: bool
    data_fim: date | None
    transacao_vinculada_id: int | None


class VincularItemPlanejadoIn(BaseModel):
    transacao_id: int
